services/web/server/src/simcore_service_webserver/projects/projects_handlers.py

```diff
@@ -53,104 +54,135 @@
-            project, nodes_map = clone_project_document(
-                source_project, forced_copy_project_id=None
+            new_project, nodes_map = clone_project_document(
+                source_project,
+                forced_copy_project_id=None,
+                clean_output_data=(copy_data == False),
             )
-            clone_data_coro = copy_data_folders_from_project(
-                request.app, source_project, project, nodes_map, user_id
+            if template_uuid:
+                # remove template access rights
+                new_project["accessRights"] = {}
+            # the project is to be hidden until the data is copied
+            hidden = copy_data
+            clone_data_coro = (
+                copy_data_folders_from_project(
+                    request.app, source_project, new_project, nodes_map, user_id
+                )
+                if copy_data
+                else None
             )
-
-            # remove template access rights
-            project["accessRights"] = {}
             # FIXME: parameterized inputs should get defaults provided by service
 
         # overrides with body
         if request.has_body:
             predefined = await request.json()
-            if project:
+            if new_project:
                 for key in OVERRIDABLE_DOCUMENT_KEYS:
                     non_null_value = predefined.get(key)
                     if non_null_value:
-                        project[key] = non_null_value
+                        new_project[key] = non_null_value
             else:
                 # TODO: take skeleton and fill instead
-                project = predefined
+                new_project = predefined
 
         # re-validate data
-        projects_api.validate_project(request.app, project)
+        await projects_api.validate_project(request.app, new_project)
 
         # update metadata (uuid, timestamps, ownership) and save
-        project = await db.add_project(
-            project,
+        new_project = await db.add_project(
+            new_project,
             user_id,
             force_as_template=as_template is not None,
             hidden=hidden,
         )
 
         # copies the project's DATA IF cloned
         if clone_data_coro:
-            await clone_data_coro
+            assert source_project  # nosec
+            if as_template:
+                # we need to lock the original study while copying the data
+                async with projects_api.lock_with_notification(
+                    request.app,
+                    source_project["uuid"],
+                    ProjectStatus.CLONING,
+                    user_id,
+                    await get_user_name(request.app, user_id),
+                ):
+
+                    await clone_data_coro
+            else:
+                await clone_data_coro
+            # unhide the project if needed since it is now complete
+            if not new_project_was_hidden_before_data_was_copied:
```

Comment:
THOUGHT: perhaps features like these should be better managed with some kind of guards implemented with context managers. Upon __enter__ we can store some fields that can be ensured upon __exit__ ...

Reply:
agreed.. here I felt a bit bad... the function did already too much before, but I was a bit reluctant of taking the time to rework it completely.